kubeadm config images list: implement structured output

bart0sh opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

This is an implementation of kubeadm structured output described in this [KEP](https://github.com/kubernetes/enhancements/pull/1054) for ```kubeadm config images list``` command. It outputs image information in TEXT, JSON, YAML, Go template and JSONPATH formats.

Example output:

```
$ kubeadm config images list
k8s.gcr.io/kube-apiserver:v1.17.0
k8s.gcr.io/kube-controller-manager:v1.17.0
k8s.gcr.io/kube-scheduler:v1.17.0
k8s.gcr.io/kube-proxy:v1.17.0
k8s.gcr.io/pause:3.1
k8s.gcr.io/etcd:3.4.3-0
k8s.gcr.io/coredns:1.6.5

$ kubeadm config images list -o json
{
    "kind": "Images",
    "apiVersion": "output.kubeadm.k8s.io/v1alpha1",
    "images": [
        "k8s.gcr.io/kube-apiserver:v1.17.0",
        "k8s.gcr.io/kube-controller-manager:v1.17.0",
        "k8s.gcr.io/kube-scheduler:v1.17.0",
        "k8s.gcr.io/kube-proxy:v1.17.0",
        "k8s.gcr.io/pause:3.1",
        "k8s.gcr.io/etcd:3.4.3-0",
        "k8s.gcr.io/coredns:1.6.5"
    ]
}

$ kubeadm config images list -o yaml
apiVersion: output.kubeadm.k8s.io/v1alpha1
images:
- k8s.gcr.io/kube-apiserver:v1.17.0
- k8s.gcr.io/kube-controller-manager:v1.17.0
- k8s.gcr.io/kube-scheduler:v1.17.0
- k8s.gcr.io/kube-proxy:v1.17.0
- k8s.gcr.io/pause:3.1
- k8s.gcr.io/etcd:3.4.3-0
- k8s.gcr.io/coredns:1.6.5
kind: Images

$ kubeadm config images list -o 'go-template={{range .images}}{{.}}{{"\n"}}{{end}}'
k8s.gcr.io/kube-apiserver:v1.17.0
k8s.gcr.io/kube-controller-manager:v1.17.0
k8s.gcr.io/kube-scheduler:v1.17.0
k8s.gcr.io/kube-proxy:v1.17.0
k8s.gcr.io/pause:3.1
k8s.gcr.io/etcd:3.4.3-0
k8s.gcr.io/coredns:1.6.5

$ kubeadm config images list -o 'jsonpath={range.images[*]}{@} {end}' | tr ' ' '\n'
k8s.gcr.io/kube-apiserver:v1.17.0
k8s.gcr.io/kube-controller-manager:v1.17.0
k8s.gcr.io/kube-scheduler:v1.17.0
k8s.gcr.io/kube-proxy:v1.17.0
k8s.gcr.io/pause:3.1
k8s.gcr.io/etcd:3.4.3-0
k8s.gcr.io/coredns:1.6.5
```

**Does this PR introduce a user-facing change?**:
```release-note
kubeadm: implemented structured output of 'kubeadm config images list' command in JSON, YAML, Go template and JsonPath formats
```

bart0sh commented:
/cc @neolit123 @fabriziopandini @rosti @akutz @kad @Klaven

bart0sh commented:
/cc @kubernetes/sig-cluster-lifecycle-pr-reviews

bart0sh commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-e2e-gce

neolit123 commented on cmd/kubeadm/app/apis/output/v1alpha1/types.go:
```diff
@@ -31,3 +31,13 @@ type BootstrapToken struct {
+
+	Name    string `json:"name"`
+	Version string `json:"version"`
+}
```
this opens the same discussion as the `upgrade plan` PR:
streaming vs object separation

my vote is for the command to have the parent `Images` structure:
```
type Images struct {
	metav1.TypeMeta `json:",inline"`

	Images []Image `json:"images"`
}

type Image struct {
	Name    string `json:"name"`
	Version string `json:"version"`
}
```

this leaves us the option to add extra contents to the Images structure in the future.
streaming is not really required for this command, and for the time being.

i'm sure @rosti will agree to this.

rosti replied:
Absolutely, that would make parsing a lot easier. In the JSON case, we end up with a bunch of objects, that are not separated in any way. Making it hard for users to parse them. In addition to that, image listing is an operation that takes a negligibly small amount of time to complete. Hence, no buffering and separate objects are needed.

bart0sh replied:
@neolit123 @rosti Thank you for the review. Looks like I should also get rid of unbuffered output in the current implementation of [token output](https://github.com/kubernetes/kubernetes/pull/78764). Please, confirm.

neolit123 replied:
does the token output looks like this?

```
BootstrapToken
---
BootstrapToken
---
```

i'm sorry but i forgot what we are doing there. it makes sense if we are fetching one token at a time, if we are fetching them all at once it makes sense to have a parent structure:

```
BootstrapTokens {
    .... []BootstrapToken
}
```

the parent structure is preferred in all cases, unless we want it unbuffered - e.g. making network calls that are slow.

neolit123 replied:
/cc @akutz 
do you have an opinion on this topic?

rosti commented on cmd/kubeadm/app/apis/output/v1alpha1/types.go:
```diff
@@ -31,3 +31,13 @@ type BootstrapToken struct {
+	metav1.TypeMeta `json:",inline"`
+
+	Name    string `json:"name"`
+	Version string `json:"version"`
```
I would prefer sticking to the term `Tag` here.

rosti commented on cmd/kubeadm/app/apis/output/v1alpha1/types.go:
```diff
@@ -31,3 +31,13 @@ type BootstrapToken struct {
+type Image struct {
+	metav1.TypeMeta `json:",inline"`
+
+	Name    string `json:"name"`
```
I guess that this includes the prefix too. Why not split them?

rosti commented on cmd/kubeadm/app/cmd/config.go:
```diff
@@ -571,11 +585,37 @@ type ImagesList struct {
 	for _, img := range imgs {
-		fmt.Fprintln(out, img)
+		nameVer := strings.SplitN(img, ":", 2)
+		outputImage := outputapiv1alpha1.Image{Name: nameVer[0], Version: nameVer[1]}
```
This is looking hacky to me. What if at some point images get pinned to digests (`image:v1.50.5` vs `image@sha256:ab82fd561234...`)?

bart0sh replied:
How you'd you suggest to parse this then?

rosti reviewed (commented):
Thanks @bart0sh !
I don't like the separate objects as opposed to image objects in a single list.
Furthermore, I am not entirely sure if it makes sense to split each image string into "prefix", "image" and "tag" (as opposed to just dumping them as strings).
Also, having the image name components ("prefix", "image" and "tag") assembled in `images.go` just to be disassembled again in `config.go` is not ideal. We should probably return them as structs from the `images.go` funcs in the first place.

bart0sh commented:
@rosti 
> Furthermore, I am not entirely sure if it makes sense to split each image string into "prefix", "image" and "tag" (as opposed to just dumping them as strings).

I think it makes sense. We don't have them splited just because we didn't have to present them in a structured way before.

> Also, having the image name components ("prefix", "image" and "tag") assembled in images.go just to be disassembled again in config.go is not ideal. We should probably return them as structs from the images.go funcs in the first place.

Agree, it's not ideal. However, changing a lot of working code just for the output is worse than that in my opinion.

neolit123 commented on cmd/kubeadm/app/cmd/config.go:
```diff
@@ -571,11 +585,45 @@ type ImagesList struct {
+	var outImages []outputapiv1alpha1.Image
 	for _, img := range imgs {
-		fmt.Fprintln(out, img)
+		splitted := strings.SplitN(img, "/", 2)
```
should be just `split`. `split` is the past tense of `split`.

neolit123 replied:
this requires error handling to check if the slice len is < 2.
[1]

bart0sh replied:
It can't be < 2 in current implementation. image is formatted by [this code](https://github.com/kubernetes/kubernetes/blob/release-1.17/cmd/kubeadm/app/images/images.go#L31).

I agree with @rosti here. It should be done differently to avoid parsing the string, which was just formatted. However, this seem to be out of scope of this PR.

bart0sh replied:
Done. Please review again.

neolit123 commented on cmd/kubeadm/app/cmd/config.go:
```diff
@@ -571,11 +585,45 @@ type ImagesList struct {
 	for _, img := range imgs {
-		fmt.Fprintln(out, img)
+		splitted := strings.SplitN(img, "/", 2)
+		nameTag := strings.SplitN(splitted[1], ":", 2)
```
[1]

neolit123 replied:
i guess @rosti's comment still applies here:
https://github.com/kubernetes/kubernetes/pull/86810#discussion_r363297497

can we use `@` to determine if this is a digest or name+tag?

then again our structure field is called `tag`, does it make sense to store digests in there at some point?

bart0sh replied:
This parsing based on [this formatting](https://github.com/kubernetes/kubernetes/blob/release-1.17/cmd/kubeadm/app/images/images.go#L32). "tag" and "prefix" names come from there.
if we want change that we'll have to do it in both places.

neolit123 replied:
let's keep it out of scope of this PR then.
@rosti WDYT?

bart0sh replied:
no splitting image name to components in the current version. Please review.

rosti commented:
@bart0sh 
> Agree, it's not ideal. However, changing a lot of working code just for the output is worse than that in my opinion.

Good programs have a clear data flow with not that many transformations in it. Certainly, we should avoid doing `A->B->C->B->D`. Moreover, such doing the reverse step (`C->B`) is usually prone to be buggy. Even if we tailor it to match precisely what we did in `B->C` someone else might misuse it in the future or update the original transformation without the reverse one.
Hence, it's a best practice to avoid those

bart0sh committed:
kubeadm config images list: update output API

bart0sh committed:
kubeadm config images list: implement structured output

Used cli-runtime API to print image info in 5 formats:

  - TEXT (identical to the current output)
  - YAML
  - JSON
  - JSONPATH
  - Go template

bart0sh committed:
kubeadm config images list: test structured output

Implemented tests for 'kubeadm config images list' structured output.

bart0sh commented:
I decided to follow this @rosti's suggestion:
> Furthermore, I am not entirely sure if it makes sense to split each image string into "prefix", "image" and "tag" (as opposed to just dumping them as strings).

And got rid of splitting image name to components. Please review.

Examples in the PR descriptions have been updated too.

bart0sh commented:
/retest

bart0sh commented:
/retest

rosti commented on cmd/kubeadm/app/cmd/config.go:
```diff
@@ -571,11 +584,43 @@ type ImagesList struct {
+func (i *ImagesList) Run(out io.Writer, printer output.Printer) error {
 	imgs := images.GetControlPlaneImages(&i.cfg.ClusterConfiguration)
+
+	var outImages []outputapiv1alpha1.Image
```
Ideally, we want to fill in stuff into an internal type. Then we can convert that to whichever version we need prior to dumping.

bart0sh replied:
now it's a slice of strings as you've suggested.

rosti reviewed (commented):
Thanks @bart0sh !
Looks that we are getting on the right track here.

bart0sh commented on cmd/kubeadm/app/apis/output/types.go:
```diff
@@ -31,3 +31,17 @@ type BootstrapToken struct {
+}
+
+// Image represents information about one image
+type Image struct {
```
Thanks for the review. done.

bart0sh commented on cmd/kubeadm/app/apis/output/v1alpha1/types.go:
```diff
@@ -31,3 +31,17 @@ type BootstrapToken struct {
+}
+
+// Image represents information about one image
+type Image struct {
```
done

bart0sh commented on cmd/kubeadm/app/cmd/config.go:
```diff
@@ -571,11 +584,43 @@ type ImagesList struct {
+
+// PrintObj is an implementation of ResourcePrinter.PrintObj for plain text output
+func (itp *imageTextPrinter) PrintObj(obj runtime.Object, writer io.Writer) error {
+	for _, img := range obj.(*outputapiv1alpha1.Images).Images {
```
done

bart0sh commented:
/retest

bart0sh commented:
/retest

neolit123 commented:
/approve
/priority important-longterm

our list of images ended up easier for us to manage using strings e.g.:
```
[
  "k8s.gcr.io/kube-apiserver:v1.16.0",
  "k8s.gcr.io/kube-controller-manager:v1.16.0",
  "k8s.gcr.io/kube-scheduler:v1.16.0",
  ...
]
```

but this delegates a parsing step to the users - e.g. separation of image and tag.
in any case this PR seems good to me, as is.

@rosti do you have more comments?

bart0sh commented:
/retest

rosti commented:
> our list of images ended up easier for us to manage using strings e.g.:
> 
> ```
> [
>   "k8s.gcr.io/kube-apiserver:v1.16.0",
>   "k8s.gcr.io/kube-controller-manager:v1.16.0",
>   "k8s.gcr.io/kube-scheduler:v1.16.0",
>   ...
> ]
> ```
> 
> but this delegates a parsing step to the users - e.g. separation of image and tag and mapping to a component.
Or easier for users who are interested in the images as they are (I can imagine - most of the users). Splitting the image in three parts was without a clear benefit to me.

> @rosti do you have more comments?
/lgtm

neolit123 commented:
> pull-kubernetes-bazel-test

unit tests seem to be failing.

bart0sh commented:
I'll check it out.

bart0sh commented:
It's caused by the recent pause version update. I'll fix this. Thanks for pointing it out.

bart0sh commented:
should be fixed now.

neolit123 commented:
thanks
/lgtm
